### PR TITLE
Story #3 Search

### LINKED ISSUE #3: [Story] Search for projects
# Value statement

As a **user**
I need to **search**
so that I can **find projects I heard about, have an alternative to tags-filtering**

# Description (Text, Scribble, Wireframe, Design)

image

-  The search bar is the prominently placed uppermost item
-  Design inspired by Googles search bar, as this will be familiar to users

# Acceptance criteria

- [x] Placeholder text vanishes after initial input (But not by tapping  the field)
- [x] Every Input is instantly visible
- [x] Cards get reordered in real time
- [x] Press on the 'X' removes input
- [x] Results are ordered in respect to their similarity with the search terms -> Compared to full text
- [x] Search and filter function together

# Tasks

- [x] Research fuzzy search &&
- [x] Implement fuzzy search 
- [x] Create 'Search' Component consisting of 🔎 +  placeholder text and the X
- [x] Sort Array of cards by tags, then accuracy to search terms, most fitting at the beginning
- [x] Testing

# Complexity
high

### PR DESCRIPTION
**Pull request by constantingoeldel**
With this update, the functionality to search for projects is added. Reviews are greatly appreciated :)
Story #3

### CODE REVIEW

File: src/stories/Project.stories.js [hunk 1]
```diff
@@ -8,28 +8,36 @@ export default {
...
-    country='AUS'
-    description='Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis tempus erat at urna semper tempus. Cras euismod semper turpis, at luctus elit dignissim non. Curabitur id arcu in justo pellentesque.'
-    image='/images/wildfire.jpg'
+    data={{
+      tags: ['Social Business', 'In-Person-Support', 'Non-Profit'],
+      title: 'Take a breather',
+      country: 'AUS',
+      description:
+        'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis tempus erat at urna semper tempus. Cras euismod semper turpis, at luctus elit dignissim non. Curabitur id arcu in justo pellentesque.',
+      image: '/images/wildfire.jpg',
```
Comment (EmilyCook321) [inline src/stories/Project.stories.js:17]: Perhaps you would like to include an "alt" attribute to your images? :)
This link may be useful: https://stackoverflow.com/questions/15471688/adding-alt-attribute-to-image-in-javascript

File: src/stories/Project.stories.js [hunk 2]
```diff
@@ -8,28 +8,36 @@ export default {
...
-    title='Take a breather'
-    country='AUS'
-    description='Lorem ipsum doloruelit dignissim non. Curabitur id arcu in justo pellentesque.'
-    image='/images/ZeroClimate.jpg'
+    data={{
+      tags: ['Legal Advice', 'Technology', 'Environment'],
+      title: 'Take a breather',
+      country: 'AUS',
+      description: 'Lorem ipsum doloruelit dignissim non. Curabitur id arcu in justo pellentesque.',
+      image: '/images/ZeroClimate.jpg',
```
Comment (EmilyCook321) [inline src/stories/Project.stories.js:28]: Same "alt atrribute" suggestion as above :)

File: src/stories/Project.stories.js [hunk 3]
```diff
@@ -8,28 +8,36 @@ export default {
...
-    country='AUS'
-    description='Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis tempus eelit dignissim non. Curabitur id arcu in justo pellentesque.'
-    image='/images/refugees.jpg'
+    data={{
+      tags: ['Social Business', 'In-Person-Support', 'People in need'],
+      title: 'Never Refuse Refuge',
+      country: 'AUS',
+      description:
+        'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis tempus eelit dignissim non. Curabitur id arcu in justo pellentesque.',
+      image: '/images/refugees.jpg',
```
Comment (EmilyCook321) [inline src/stories/Project.stories.js:40]: Same "alt atrribute" suggestion as above :)

File: src/utils.test.js [hunk 4]
```diff
@@ -82,3 +82,35 @@ describe('test sorting function', () => {
...
+    // expect(filterBySearch(undefined, undefined)).toBeDefined()
+  })
+  it('Filters countries correctly', () => {
+    expect(filterBySearch({ target: { value: 'LIB' } }, mockData)).toEqual([
+      {
+        title: 'Never Refuse Refuge',
+        country: 'LIB',
+        description:
+          'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis tempus erat at urna semper tempus. Cras turpis, at luctus elit',
+        image: '/images/refugees.jpg',
```
Comment (EmilyCook321) [inline src/utils.test.js:98]: Same "alt atrribute" suggestion as above :)

**Review by EmilyCook321 — COMMENTED** [review]
Overall, looks good- just a suggestion on adding alt to the images included for improved accessibility.

File: src/stories/Project.stories.js (hunk 1 above)
Comment (constantingoeldel) [reply src/stories/Project.stories.js:17]: Currently, all images get an empty string as the alt-attribute as I have no way of knowing what they are about due to them being selected at random from unsplash. Potentially I will let the user add a alt when creating the project

**Comment by karstenlenz** [comment]
QA feedback:

- The search input has the same style as the card component. I would suggest giving it a separate styling so it can immediately be recognized.
- It is not immediately clear, what you can search for. Maybe you can add some hint (e.g. "you can search for country, title or description"
- The search is case sensitive.  "take a breather" won't find the "Take a breather" card. 
- some search results don't match the search query. I searched for "social business", and got the "ID | 4th Floor, The" card which neither has "social" or "business" in the text.  
- If you have filters activated it may confuse a user that a search won't return a result. May be add a hint such as "The result is filtered by tags. Some search results may be hidden"

**Pull request merged**